### PR TITLE
Enable PublicApiAnalyzers

### PR DESCRIPTION
**Pull request by westin-m**
See: https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md

Include internals for assemblies that are marked with InternalsVisibleTo

### CODE REVIEW

File: benchmark/Directory.Build.props [hunk 1]
```diff
@@ -20,4 +20,9 @@
   <ItemGroup>
     <AdditionalFiles Include="$(MSBuildThisFileDirectory)..\BannedSymbols.txt" />
   </ItemGroup>
+
+  <PropertyGroup>
+    <!--RS0016: Add public types and members to the declared API-->
```
Comment (jmprieur) [inline benchmark/Directory.Build.props:25]: why is this warning removed?

File: src/Microsoft.Identity.Web.Azure/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+﻿Microsoft.Identity.Web.TokenAcquirerAppTokenCredential
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web.Azure/PublicAPI.Unshipped.txt:1]:
```suggestion
﻿﻿#nullable enable
﻿﻿Microsoft.Identity.Web.TokenAcquirerAppTokenCredential
```

File: src/Microsoft.Identity.Web.Certificate/PublicAPI.Unshipped.txt [hunk 3]
```diff
@@ -0,0 +1,42 @@
+﻿Microsoft.Identity.Web.CertificateDescription
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web.Certificate/PublicAPI.Unshipped.txt:1]:
```suggestion
﻿#nullable enable
﻿﻿Microsoft.Identity.Web.CertificateDescription
```

File: src/Microsoft.Identity.Web.Certificateless/PublicAPI.Unshipped.txt [hunk 4]
```diff
@@ -0,0 +1,24 @@
+﻿abstract Microsoft.Identity.Web.ClientAssertionProviderBase.GetClientAssertionAsync(Microsoft.Identity.Client.AssertionRequestOptions assertionRequestOptions) -> System.Threading.Tasks.Task<Microsoft.Identity.Web.ClientAssertion>
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web.Certificateless/PublicAPI.Unshipped.txt:1]:
```suggestion
﻿#nullable enable
﻿﻿abstract Microsoft.Identity.Web.ClientAssertionProviderBase.GetClientAssertionAsync(Microsoft.Identity.Client.AssertionRequestOptions assertionRequestOptions) -> System.Threading.Tasks.Task<Microsoft.Identity.Web.ClientAssertion>
```

Comment (jmprieur) [reply src/Microsoft.Identity.Web.Certificateless/PublicAPI.Unshipped.txt:1]: all the Public API files (shipped and unshipped) need to start with 
```
#nullable enable
```

**Comment by JoshLozensky** [comment]
Do we need to enable the analyzers on our tests?

**Comment by JoshLozensky** [comment]
Is there a process for moving things from unshipped to shipped?

**Comment by ciaozhang** [comment]
> Is there a process for moving things from unshipped to shipped?

We have guidance in MISE:
https://identitydivision.visualstudio.com/DevEx/_git/MISE?path=/internaldocs/Public_API_Analyzer.md&_a=preview

File: .editorconfig [hunk 5]
```diff
@@ -141,6 +141,19 @@ csharp_space_between_square_brackets = false
 # RS0030: Do not used banned APIs
 dotnet_diagnostic.RS0030.severity = error
 
+# RS0037: Enable tracking of nullability of reference types in the declared API
+dotnet_diagnostic.RS0037.severity = silent
```
Comment (keegan-caruso) [inline .editorconfig:145]: since the api files are using nullables, this shouldn't be silent.

**Comment by keegan-caruso** [comment]
mark-shipped should look like this: https://github.com/dotnet/aspnetcore/blob/2287c782562158ebe86b438ab9792db244764b48/eng/scripts/mark-shipped.ps1

**Comment by JoshLozensky** [comment]
> It is a step for release process. We have guidance in MISE:

What is the impact if we get this wrong? As a partially manual process we will probably get it wrong at some point. Like if we revert a change but forget to move it back to unreleased, or miss adding changes to the framework file, miss a forwarded type, etc...

**Comment by westin-m** [comment]
> Do we need to be enabling the analyzers on our tests? Their methods aren't packed into our libraries.

No, those files slipped past me after being generated automatically.

File: .editorconfig [hunk 6]
```diff
@@ -64,7 +64,7 @@ dotnet_naming_style.pascal_case_style.capitalization = pascal_case
 # static fields should have s_ prefix
 dotnet_naming_rule.static_fields_should_have_prefix.severity = suggestion
 dotnet_naming_rule.static_fields_should_have_prefix.symbols  = static_fields
-dotnet_naming_rule.static_fields_should_have_prefix.style    = static_prefix_style
+dotnet_naming_rule.static_fields_should_have_prefix.style = static_prefix_style
```
Comment (jennyf19) [inline .editorconfig:67]: for future reference, would be good to have the style updates in a separate PR. less to review

**Review by jennyf19 — APPROVED** [review]

File: .editorconfig (hunk 6 above)
Comment (westin-m) [reply .editorconfig:67]: Sorry about that, VS did these without asking

**Review by JoshLozensky — APPROVED** [review]

**Pull request merged**